docs: Identity Center Okta to Teleport migration guide (#51861)
gravitational/teleport: docs/pages/admin-guides/management/guides/migrating-iam-ic-from-okta-to-teleport.mdx

---
title: Migrating AWS Identity Center from Okta to Teleport
description: A guide for gradually migrating an Identity Center instance from Okta control to Teleport control
tocDepth: 3
---

## Overview

The goal of this guide is to alow you to transition from Okta managing
Identity Center users and groups to Teleport management, including having
Teleport control individual user and group Account Assignments.

> By default, Teleport wants to control all user and group provisioning into 

Reviewer: ```suggestion
By default, Teleport controls all user and group provisioning into 
```

 both applications.
</Admonition>

You will need the Teleport SAML Connector Name and Okta SAML App ID in the next
step.

## Step 2: Install the Teleport Okta integration

We will be using a very limited subset of the Teleport Okta integration in this 
deployment, disabling all features except periodic user synchronization. This
configuration is not supported by the normal installation UI, so we will have to
use `tctl` to install it:

```console
$ ./tctl plugins install okta \
    --org ${OKTA_ORG_URL} \
    --saml-connector ${TELEPORT_SAML_CONNECTOR_NAME} \
    --app-id ${OKTA_SAML_APP_ID} \
    --api-token ${OKTA_API_TOKEN} \
    --no-scim \
    --no-accesslist-sync \
    --no-appgroup-sync
```

This will install the Okta integration and start the User sync service. You can
monitor the state of the Okta integration in the Teleport Integrations UI.

## Step 3: Wait for user sync

To make sure everything is working, wait until the first Okta to Teleport user
sync has occurred. You can verify this by either
 - refreshing the user page and finding your Okta users,
 - checking the Okta integration status page, or

>  - watching the Teleport logs

Reviewer: Nit: I'd remove this one in favor of the previous two more user-friendly options. Plus, it won't suit Cloud customers.
```suggestion
```

Once your Okta users are imported into Teleport, you can progress to the next

> step:

Reviewer: ```suggestion
step.
```

## Step 4: Install the Teleport Identity Center Integration.

Again, we need to install the plugin using `tctl`. 

```console
$ ./tctl plugins install awsic \
    --arn ${IDENTITY_CENTER_INSTANCE_ARN} \
    --region ${IDENTITY_CENTER_INSTANCE_REGION} \
    --use-system-credentials \
    --url ${IDENTITY_CENTER_SCIM_BASE_URL} \
    --token ${IDENTITY_CENTER_SCIM_BEARER_TOKEN} \
    --default-owner ${TELEPORT_ACCESS_LIST_DEFAULT_OWNER} \
    --user-origin okta \
    --account-name ${ACCOUNT_NAME_ALLOW_FILTER} \

>     --group-name ${GROUP_NAME_ALLOW_FILTER}

Reviewer: A logical question that arises, how can these filters be updated after the integration has been created? I didn't see a mention of this later in the guide.

By default, Teleport will take control of Account Assignments for all AWS Accounts
managed by Identity Center. You can create an allow-list of AWS Accounts to
import with the  `--account-name` and `--account-id` install options. 

The `--account-name` filters work like the `--group-name` filters above. The 
`--account-id` filters specify a literal AWS Account ID.

Teleport will not create or delete Account Assignments on AWS accounts outside
of its allow-list.

## Step 5: Migrate Account Assignments

We're now at the midway point, and ready to migrate Account Assignments from the
Okta-managed groups into new Teleport-managed Access Lists. To migrate groups,
create a new Access List in Teleport (taking care not to use the same name as
the existing Okta-managed Group) and create the appropriate memberships and 
Account Assignments.

Account Assignments can be created on an Access List by assigning it the Account
Assignment roles created by the Identity Center integration, assigning it a 
custom Teleport role that specifies a specific combination of access, or a
combination of each.

For more information, see the [Identity Center Integration guide](./aws-iam-identity-center.mdx)

> ## Step 6: Retire Okta group provisioning

Reviewer: Should probably also include a section on how to unenroll the Teleport integration side of this as well in case someone decides they don't want to switch yet.

 - Teleport controls IC Group Account Assignments for the Identity Center groups under its control
 - Teleport controls direct IC User Account Assignments for the Identity Center users under its control

![Migration Mid Point](../../../../img/identity-center/ic-migrate-mid.png)

#### Ending point

Once Teleport is in control of all the users and groups you want to provision
into Identity Center, the Okta Provisioning and Push Groups functions can be
retired, leaving Okta only providing SSO login.

![Migration End Point](../../../../img/identity-center/ic-migrate-end.png)

## Prerequisites

 - A working Teleport cluster
 - An AWS role configured as per the [Identity Center guide](./aws-iam-identity-center.mdx)
   for the integration to use
 - AWS credentials configured for the Teleport auth process to pick up and use 
   (e.g. as environment vars, system profiles, etc)
 - An Okta API token, with the following privileges:
   - View users and their details
   - View groups and their details
   - View applications and their details
 - The AWS Identity Center ARN, AWS region, SCIM base address and SCIM bearer token

> ## Step 1: Install Okta SAML connector

Reviewer 2: Try to use the `Step 1/N` format for consistency with other guides.

Install Okta SAML connector into Teleport as per the Teleport 
[Okta as an SSO provider](../../access-controls/sso/okta.mdx) guide.

<Admonition type="note">
 Ideally, both Identity Center and Teleport will use the same Okta SAML application
 for their signin, which guarantees that both Okta and Teleport will view the same 
 user set. 

 If this is not possible, the most flexible approach we've found is having an
 an Okta Group for those users that should have IC access, and then assigning
 both the Okta Identity Center App and the Okta SAML App used for Teleport SAML
 connector to that group. This ensures that the same user set is visible across
 both applications.
</Admonition>

You will need the Teleport SAML Connector Name and Okta SAML App ID in the next
step.

## Step 2: Install the Teleport Okta integration

We will be using a very limited subset of the Teleport Okta integration in this 
deployment, disabling all features except periodic user synchronization. This

> configuration is not supported by the normal installation UI, so we will have to

Reviewer: ```suggestion
configuration is not currently supported by the normal installation UI, so we will have to
```

use `tctl` to install it:

```console

> $ ./tctl plugins install okta \

Reviewer 2: ```suggestion
$ tctl plugins install okta \
```

    --org ${OKTA_ORG_URL} \
    --saml-connector ${TELEPORT_SAML_CONNECTOR_NAME} \
    --app-id ${OKTA_SAML_APP_ID} \
    --api-token ${OKTA_API_TOKEN} \
    --no-scim \
    --no-accesslist-sync \
    --no-appgroup-sync
```

This will install the Okta integration and start the User sync service. You can
monitor the state of the Okta integration in the Teleport Integrations UI.

## Step 3: Wait for user sync

To make sure everything is working, wait until the first Okta to Teleport user
sync has occurred. You can verify this by either
 - refreshing the user page and finding your Okta users,

>  - checking the Okta integration status page, or

Reviewer 2: or what?

Once your Okta users are imported into Teleport, you can progress to the next
step.

## Step 4: Install the Teleport Identity Center Integration.

Again, we need to install the plugin using `tctl`. 

```console

> $ ./tctl plugins install awsic \

Reviewer 2: ```suggestion
$ tctl plugins install awsic \
```

    --arn ${IDENTITY_CENTER_INSTANCE_ARN} \
    --region ${IDENTITY_CENTER_INSTANCE_REGION} \
    --use-system-credentials \
    --url ${IDENTITY_CENTER_SCIM_BASE_URL} \
    --token ${IDENTITY_CENTER_SCIM_BEARER_TOKEN} \
    --default-owner ${TELEPORT_ACCESS_LIST_DEFAULT_OWNER} \
    --user-origin okta \
    --account-name ${ACCOUNT_NAME_ALLOW_FILTER} \
    --group-name ${GROUP_NAME_ALLOW_FILTER}
```

During the installation process, Teleport will import all of the Identity Center 
groups that match its allow list (or all of them, if no allow list is defined)
and create matching Access Lists, preserving the group membership and account
assignments.

<Admonition type="warning">
Individual User Account Assignments will _not_ be preserved during import. You
will need to ensure these are preserved manually, or converted to group assignments
prior to installation.
</Admonition>

### Group import control

> The Group import allow-list is controlled by the `--group-name` option. You can

Reviewer 2: ```suggestion
The Group import allow list is controlled by the `--group-name` option. You can
```

specify multiple filters and a Group will be imported if matches _any_ of the
supplied filters. Filters can be either literal names, globbed names or Go
compatible regular expressions. To treat a filter as a regular expression, 
enclose it in a leading `^` and trailing `$`.

Example filters:
 - `administrators`: The literal "administrators" group
 - `site-*`: Any group with the prefix `site-`
 - `^(?:[^a]|a[^w]|aw[^s]|aws[^\-]).*$`: Any group that does _not_ have the prefix `aws-`

<Admonition type="note">
Ensure that there is no overlap between the groups imported to Teleport and the
groups you want Okta to maintain control over.
</Admonition>

> <Admonition type="warning">

Reviewer 2: Stacking admonitions like this creates a lot of noise. I would prefer more plain text and only use the admonition when there's something you really need to highlight.

the existing Okta-managed Group) and create the appropriate memberships and 
Account Assignments.

Account Assignments can be created on an Access List by assigning it the Account
Assignment roles created by the Identity Center integration, assigning it a 
custom Teleport role that specifies a specific combination of access, or a
combination of each.

For more information, see the [Identity Center integration guide](./aws-iam-identity-center.mdx)

## Step 6: Expanding Teleport integration scope

Once you are satisfied with the way Teleport is handling the initial set of
imported AWS resources, you can expand the scope of the Identity Center
integration by editing the plugin import features. 

### With `tctl edit`

<Admonition type="warning">
This currently involves manually editing the Identity Center's integration 
resource using `tctl`, which is a dangerous operation. Please ensure you take a
backup of the plugin resource so you can roll back if necessary.

A guided editing workflow is currently under development.
</Admonition>

You can expand the scope of the Teleport Identity Center integration by editing
the integration's plugin resource with `tctl`.

```console

> ./tctl edit plugins/aws-identity-center

Reviewer: ```suggestion
$ tctl edit plugins/aws-identity-center
```

```

The plugin resource is a YAML document that looks something like this:

```YAML
kind: plugin
version: v1
metadata:
  labels:
    teleport.dev/hosted-plugin: "true"
  name: aws-identity-center
spec:
  Settings:
    aws_ic:
      # Account import filters. An absent or empty list of filters implies "manage all AWS accounts"
      aws_accounts_filters:
      - Include:
          id: "637423191929"
      - Include:
          id: "730335414865"
      - Include:
          id: "058264527036"
      - Include:
          name_regex: ^Staging-.*$

      # User provisioning filters. An absent or empty list of filters implies "provision all users to AWS"
      user_sync_filters:
      - labels:
          teleport.dev/origin: okta

      # Group import filters. See notes below. 
      group_sync_filters:
      - Include:
          name_regex: '^Group #00\d+$'

      access_list_default_owners:
      - admin
      arn: arn:aws:sso:::instance/ssoins-722326ecc902a06a
      credentials_source: 2
      integration_name: aws-identity-center
      provisioning_spec:
        base_url: https://scim.us-east-1.amazonaws.com/f3v9c6bc2ca-b104-4571-b669-f2eba522efe8/scim/v2
      region: us-east-1
```

> You can add or remove filters to the various filters. Once you save and quit the

Reviewer: "filters to the various filters" reads strange

Author: And this is why proofreading exists.

editor, `tctl` will replace the existing resource with your updated version. This

> will automatically restart the Identity Center Integration with the ne2 filters.

Reviewer: ```suggestion
will automatically restart the Identity Center Integration with the new filters.
```

Install Okta SAML connector into Teleport as per the Teleport 
[Okta as an SSO provider](../../access-controls/sso/okta.mdx) guide.

<Admonition type="note">
 For the integration to function properly, both AWS Identity Center and Teleport must view the
 same user set which can be achieved by using the same Okta SAML application for both Identity
 Center and Teleport SAML connector.

 If this is not possible, the most flexible approach we've found is having an
 an Okta Group for those users that should have IC access, and then assigning
 both the Okta Identity Center App and the Okta SAML App used for Teleport SAML
 connector to that group. This ensures that the same user set is visible across
 both applications.
</Admonition>

You will need the Teleport SAML Connector Name and Okta SAML App ID in the next
step.

## Step 2/7. Install the Teleport Okta integration

We will be using a very limited subset of the Teleport Okta integration in this 
deployment, disabling all features except periodic user synchronization. This
configuration is not currently supported by the normal installation UI, so we
will have to use `tctl` to install it:

```console

> $ tctl plugins install okta \

Reviewer: Is there a command reference for this command and "install awsic" used below? I think we should have it to explain all the flags: https://goteleport.com/docs/reference/cli/tctl/.

Here (and below where "install awsic" is used), it might be helpful to briefly explain what effect these flags have as well.

Author: Added in #52878

    --org ${OKTA_ORG_URL} \
    --saml-connector ${TELEPORT_SAML_CONNECTOR_NAME} \
    --app-id ${OKTA_SAML_APP_ID} \
    --api-token ${OKTA_API_TOKEN} \
    --no-scim \
    --no-accesslist-sync \
    --no-appgroup-sync
```

This will install the Okta integration and start the User sync service. You can
monitor the state of the Okta integration in the Teleport Integrations UI.

## Step 3/7. Wait for user sync

To make sure everything is working, wait until the first Okta to Teleport user
sync has occurred. You can verify this by either
 - refreshing the user page and finding your Okta users, or
 - checking the Okta integration status page

Once your Okta users are imported into Teleport, you can progress to the next
step.

## Step 4/7. Install the Teleport Identity Center Integration.

Again, we need to install the plugin using `tctl`. 

```console
$ tctl plugins install awsic \
    --arn ${IDENTITY_CENTER_INSTANCE_ARN} \
    --region ${IDENTITY_CENTER_INSTANCE_REGION} \

>     --use-system-credentials \

Reviewer 3: Can we add more details on use-system-credent? 
Like provide hint that it refers to AWS credentials. Specifically, it should clarify that use-system-credent will allow to load AWS credentials from the local environment in Teleport Auth to authenticate with the AWS IC API.

import at installation time, and there is no user accessible method of re-triggering
the import. The only way to apply new group filters is to uninstall and re-install
the plugin. 

A later release of Teleport will re-trigger a group import when it detects a
change in the group import filters.
</Admonition>

### Uninstall/reinstall

A second alternative is to [uninstall](#deleting-the-identity-center-integration)
and re-install the integration with the new parameters. Be warned that this may 
result in user access disruptions while the integration is offline.

## Step 7/7. Retire Okta group provisioning

Once you are satisfied that an AWS Identity Center group has been migrated to 
Teleport control, you can remove the corresponding push Group from the Okta
Identity Center integration.

## Deleting the Identity Center integration

If you decide not to switch over to Teleport you can delete the Identity Center
integration in two ways.

Deleting the integration will automatically remove all of the Teleport resources 
used by the integration to manage its own state, as well as all of the Teleport
roles created by the integration used to grant AWS Identity Center account

> assignments and Access Lists imported from AWS Identity Center Groups.

Reviewer: Is this true? Do we actually remove all imported access lists as well? I think we keep them for Okta integration, right? Just making sure this is accurate (and does the UI warn user about what will be removed? it definitely should cc @flyinghermit)

Also, what happens with access lists with IC-assigned-roles you created yourself that get pushed to IC? Those aren't removed, correct? This should be clarified here as well.

Author: Can confirm that this is the current behaviour. No warnings in `tctl plugins delete` at the moment.

Author: User created roles and access lists will not be deleted. If a user created access list refers to a plugin-created role, it will be left with a dangling role reference once the integration is deleted, and will need manual cleanup.

I have a bunch of `// TODO:`s in various places for walking User and Access List role assignments and clean them up when deleting plugin-created roles, but nothing actually implemented for it. 

Will add some notes about the above to the doc.

Reviewer: Can you create a ticket to add confirmation/warning to tctl plugins delete? Maybe @flyinghermit can address it at the same time we add a similar warning to the UI.

Reviewer 4: Ticket added https://github.com/gravitational/teleport.e/issues/6165

---
title: Migrating AWS Identity Center from Okta to Teleport
description: A guide for gradually migrating an Identity Center instance from Okta control to Teleport control
tocDepth: 3
---

> The goal of this guide is to allow you to perform a staged migration from

Reviewer 5: Moving to active voice to keep consistent:

This guide helps you migrate AWS Identity Center users and groups from Okta to Teleport in stages while keeping Okta as your AWS SSO provider

Okta-managed AWS Identity Center users & groups, to having them managed by Teleport
while preserving Okta as your AWS SSO provider.

> This approach lets you get set up with a minimal Teleport AWS Identity Center

Reviewer 5: This approach sets up a minimal Teleport AWS Identity Center integration that you can expand as you become more familiar with Teleport. It also provides deployment options unavailable in a standard Teleport Identity Center setup, such as deploying Teleport in a private network.

integration that you can expand as you gain expertise and confidence with
Teleport, as well as offering deployment options that are not available with a 
"standard" Teleport Identity Center configuration, such as having Teleport deployed
in a private network.

> You can also use this setup if you don't want to migrate fully to using Teleport

Reviewer 5: Use this setup if you prefer not to fully migrate to Teleport as an Identity Provider for AWS Identity Center. By following this guide, you can let Teleport manage user and group AWS account assignments while keeping Okta as AWS Identity Center’s IdP.

as an Identity Provider for AWS Identity Center. Following the steps in this
guide will allow you to have Teleport control user and group AWS account
Assignments, while preserving Okta as AWS Identity Center's IdP.

## How it works

By default, the Teleport AWS Identity Center integration:
- controls all user and group provisioning into Identity Center, and 
- becomes the SSO provider for AWS. 

This is fine for most deployments, but can cause difficulties when deploying into
a sophisticated, Okta-managed AWS Identity Center configuration. 

> By contrast, the deployment in this guide will transition transition control

Reviewer 4: ```suggestion
By contrast, the deployment in this guide will transition control
```

> from Okta to Teleport in a staged fashion.

Reviewer 4: nit: elaborate on what is expected
```suggestion
from Okta to Teleport in a staged fashion where user identities are managed by Okta but account permission assignment is managed by Teleport.
```

### Migration Path

#### Starting Point

This is our starting configuration: Okta as sole Identity Source for Identity Center.

![Migration Starting Point](../../../../img/identity-center/ic-migrate-start.png)

#### Midway

This is the point where you can start migrating control of users and groups from
Okta to Teleport. At this stage:

 - Okta provides SSO login for Identity Center
 - Okta manages a subset of the Identity Center group membership (selected via Push Groups)
 - Okta and Teleport both control user provisioning via SCIM
 - Teleport manages a second subset of Identity Center groups (selected via group filters during plugin installation)
 - Teleport controls IC Group Account Assignments for the Identity Center groups under its control

>  - Teleport controls direct IC User Account Assignments for the Identity Center users under its control

Reviewer 5: Teleport controls direct IC user account Assignments for the Identity Center users under its control
Comment

(lowercase, following our previous standard in other docs, like in /okta/hosted-guide)

![Migration Mid Point](../../../../img/identity-center/ic-migrate-mid.png)

#### Ending point

Once Teleport is in control of all the users and groups you want to provision
into Identity Center, the Okta Provisioning and Push Groups functions can be
retired, leaving Okta only providing SSO login.

![Migration End Point](../../../../img/identity-center/ic-migrate-end.png)

## Prerequisites

 - A working Teleport cluster

>  - An AWS role configured as per the [Identity Center guide](./aws-iam-identity-center.mdx)

Reviewer 4: I think we should scope down the sub-heading link so reader can dive straight into it instead of reading all of that doc. E.g. `(./awsxxxx.mdx#sub-heading)`

 same user set which can be achieved by using the same Okta SAML application for both Identity
 Center and Teleport SAML connector.

 If this is not possible, the most flexible approach we've found is having an
 an Okta Group for those users that should have IC access, and then assigning
 both the Okta Identity Center App and the Okta SAML App used for Teleport SAML
 connector to that group. This ensures that the same user set is visible across
 both applications.
</Admonition>

You will need the Teleport SAML Connector Name and Okta SAML App ID in the next
step.

## Step 2/7. Install the Teleport Okta integration

We will be using a very limited subset of the Teleport Okta integration in this 
deployment, disabling all features except periodic user synchronization. This
configuration is not currently supported by the normal installation UI, so we
will have to use `tctl` to install it:

```console
$ tctl plugins install okta \
    --org ${OKTA_ORG_URL} \
    --saml-connector ${TELEPORT_SAML_CONNECTOR_NAME} \
    --app-id ${OKTA_SAML_APP_ID} \
    --api-token ${OKTA_API_TOKEN} \
    --no-scim \
    --no-accesslist-sync \
    --no-appgroup-sync
```

> This will install the Okta integration and start the User sync service with a

Reviewer 5: This will install the Okta integration and start the user sync service with a
configuration that:
- Imports Okta users assigned to the Okta App `${OKTA_SAML_APP_ID}`, and keeps them synced with the upstream Okta organization.
- Does *not* expose a SCIM service.
- Does *not* attempt to sync or manage any other resources from Okta.

```

This will install the Teleport AWS Identity Center integration with a
Teleport configuration that:
- controls the AWS Identity Center instance indicated by `--instance-arn`.
- uses the system AWS credentials to authenticate with AWS (from `--use-system-credentials`)
- manages account assignments all AWS accounts that match the `${ACCOUNT_NAME_ALLOW_FILTER}`.
- provisions all users imported from Okta into AWS Identity Center (from the `--user-origin okta` flag).
- only imports all groups matching `${GROUP_NAME_ALLOW_FILTER}` into Teleport as Access
  Lists, with `${TELEPORT_ACCESS_LIST_DEFAULT_OWNER}` as the owner. 

Note that the `tctl` installer currently only supports installations using
system-level AWS credentials with `--use-system-credentials`.

You can change the AWS Account, Group and User filters later by following the
instructions in [Step 6](#step-67-expanding-teleport-integration-scope).

During the installation process, Teleport will import all of the Identity Center 
groups that match its allow list (or all of them, if no allow list is defined)
and create matching Access Lists, preserving the group membership and account
assignments.

<Admonition type="warning">

> Individual User Account Assignments will _not_ be preserved during import. You

Reviewer 5: Individual user account Assignments will _not_ be preserved during import.

will need to ensure these are preserved manually, or converted to group assignments
prior to installation.
</Admonition>

### Group import control

The Group import allow list is controlled by the `--group-name` option. You can
specify multiple filters and a Group will be imported if matches _any_ of the
supplied filters. Filters can be either literal names, globbed names or Go
compatible regular expressions. To treat a filter as a regular expression, 
enclose it in a leading `^` and trailing `$`.

Example filters:
 - `administrators`: The literal "administrators" group
 - `site-*`: Any group with the prefix `site-`

>  - `^(?:[^a]|a[^w]|aw[^s]|aws[^\-]).*$`: Any group that does _not_ have the prefix `aws-`

Reviewer 5: ***not***

You can add or remove filters to the various filter sets. Once you save and quit
the editor, `tctl` will replace the existing resource with your updated version.
This will automatically restart the Identity Center Integration with the new
filters.

<Admonition type="note">
As of Teleport 17.3, the `group_sync_filters` are only applied during the initial
import at installation time, and there is no user accessible method of re-triggering
the import. The only way to apply new group filters is to uninstall and re-install
the plugin. 

A later release of Teleport will re-trigger a group import when it detects a
change in the group import filters.
</Admonition>

### Uninstall/reinstall

A second alternative is to [uninstall](#deleting-the-identity-center-integration)
and re-install the integration with the new parameters. Be warned that this may 
result in user access disruptions while the integration is offline.

## Step 7/7. Retire Okta group provisioning

Once you are satisfied that an AWS Identity Center group has been migrated to 
Teleport control, you can remove the corresponding push Group from the Okta
Identity Center integration.

## Deleting the Identity Center integration

> If you decide not to switch over to Teleport you can delete the Identity Center

Reviewer 5: If you decide not to switch to Teleport, you can delete the Identity Center integration in two ways.

Deleting the integration automatically removes all Teleport resources it used to manage its state, including:

- Teleport roles created for AWS Identity Center account assignments
- Access Lists imported from AWS Identity Center groups

However, user-created resources remain intact, including:

- Access Lists you created
- Roles you manually configured for account assignments

If an Access List grants permissions to a now-deleted integration role, or if a user has a deleted role assigned directly, you must manually remove those references.
